Migrate from DocumentEditor to Editor (for atomic commands) (Resolves #1125)

matthew-carroll opened the pull request:
Migrate from DocumentEditor to Editor (for atomic commands) (Resolves #1125)

The goal of this PR is to migrate to the new `Editor` API with its command pipeline, without introducing any bugs, or changing any behaviors.

I'd like for reviewers to put this PR through all possible paces and report reproduction steps for any bugs that are found.

This PR is not focused on API design. We'll continue to iterate this API as we move forward with future PRs.

Here's a guide for the concepts in this migration: https://github.com/superlistapp/super_editor/wiki/How-it-works:-The-Editor

Here's a migration guide: https://github.com/superlistapp/super_editor/wiki/Migration-Guide:-Super-Editor--=-0.2.x-to-Super-Editor-0.3.0

Please let me know as you find things that aren't in the migration guide, so I can add them.

matthew-carroll commented:
@Jethro87 @jmatth - it would be great if you could attempt a migration to this PR branch and report back with any bugs

matthew-carroll commented:
@Jethro87 @jmatth - it would be great if you could attempt a migration to this PR branch and report back with any bugs

jmatth commented:
@matthew-carroll I'll try to make time to pull the latest changes into our fork by the end of this week. Currently working on a big release I want to get done before anything else

Jethro87 commented:
@matthew-carroll I've made the migration and have just begun to test it. 

I'm noticing some issues with the formatting toolbar and selection not updating properly, particularly for headers. Check out the video below. I couldn't get the touch events to appear in the recording, unfortunately, but I'm clicking the header buttons constantly. It's only when I double-tap the text itself that the changes are applied.

https://github.com/superlistapp/super_editor/assets/6467808/475aa190-d3a4-4aad-97bd-6cde31182b1f

matthew-carroll commented:
@Jethro87 it looks like the issue is that the user can update the document selection, but we don't make any changes to the document composing region. This leads to an attempt to create an IME value with the composing region for the previously selected node, which then fails because the selected content no longer includes that node.

I think the fix might be to simply clear out the composing region whenever the user changes the selection. I'm looking into it.

matthew-carroll commented:
I pushed a change that will hopefully fix the selection issue.

I incorporated composing region changes into selection changes. Whenever a selection change is requested, and no composing region is specified, the composing region is cleared out. In theory, if the IME wants a composing region for a new selection, the IME will immediately send us the composing region that it wants. So I think clearing the composing region on selection change is reasonable.

Jethro87 commented:
@matthew-carroll I just did a quick test and nothing seems to have changed. I'll try to give the editor a more thorough test tomorrow.

matthew-carroll commented:
@Jethro87 I think I may have misunderstood what your video was showing. I think I discovered a different bug and thought it was the one you were describing.

I'll take another look at that format bar.

matthew-carroll commented:
@Jethro87 I think I've now fixed the problem in your video. We were editing the block type directly, without going through the editor pipeline, which meant listeners weren't being notified, and therefore styles weren't being re-applied.

Jethro87 commented:
@matthew-carroll It looks like that solved the issue.

I found an unrelated bug. Lines 165-167 in `default_document_editor.dart` should be changed:

```
// Current (bug)
  (request) => request is ConvertListItemToParagraphRequest //
      ? ConvertListItemToParagraphCommand(nodeId: request.nodeId)
      : null,

// Fixed
  (request) => request is ConvertListItemToParagraphRequest //
      ? ConvertListItemToParagraphCommand(nodeId: request.nodeId, paragraphMetadata: request.paragraphMetadata)
      : null,

```

When converting from a list item to a paragraph node, the metadata wasn't being passed. This caused an attempted ListItem -> Header3 attribution to end up as ListItem -> null attribution.

I'll keep testing.

matthew-carroll commented:
Thanks @Jethro87 - I pushed a fix

angelosilvestre commented:
@matthew-carroll I noticed a bug when typing links in the demo app:

1. Place the caret at the end of the paragraph "SuperEditor is a toolkit..."
2. Press enter
3. Type www.google.com
4. Press space
5. Press backspace twice
6. Press space again

The editor crashes with the following exception:

```
════════ Exception caught by services library ══════════════════════════════════
The following IncompatibleOverlappingAttributionsException was thrown during method call TextInputClient.updateEditingStateWithDeltas:
Tried to insert attribution ([LinkAttribution]: www.google.co) over a conflicting existing attribution ([LinkAttribution]: www.google.com). The overlap began at index 0

When the exception was thrown, this was the stack
#0      AttributedSpans.addAttribution
#1      AttributedText.addAttribution
#2      LinkifyReaction.react
#3      Editor._reactToChanges
#4      Editor.execute
#5      CommonEditorOperations.insertPlainText
#6      TextDeltasDocumentEditor.insert
#7      TextDeltasDocumentEditor._applyInsertion
#8      TextDeltasDocumentEditor.applyDeltas
#9      DocumentImeInputClient.updateEditingValueWithDeltas
#10     DeltaTextInputClientDecorator.updateEditingValueWithDeltas
#11     TextInput._handleTextInputInvocation
#12     TextInput._loudlyHandleTextInputInvocation
#13     MethodChannel._handleAsMethodCall
#14     MethodChannel.setMethodCallHandler.<anonymous closure>
#15     _DefaultBinaryMessenger.setMessageHandler.<anonymous closure>
#16     _invoke2 (dart:ui/hooks.dart:174:13)
#17     _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:40:5)
#18     _Channel.push (dart:ui/channel_buffers.dart:130:31)
#19     ChannelBuffers.push (dart:ui/channel_buffers.dart:326:17)
#20     PlatformDispatcher._dispatchPlatformMessage (dart:ui/platform_dispatcher.dart:686:22)
#21     _dispatchPlatformMessage (dart:ui/hooks.dart:86:31)
call: MethodCall(TextInputClient.updateEditingStateWithDeltas, [1, {deltas: [{oldText: www.google.co, deltaText:  , deltaStart: 13, deltaEnd: 13, selectionAffinity: TextAffinity.downstream, selectionBase: 14, selectionExtent: 14, selectionIsDirectional: false, composingBase: -1, composingExtent: -1}]}])
════════════════════════════════════════════════════════════════════════════════
```

angelosilvestre commented:
Steps to reproduce:

1. Run the demo app on Android
2. Place the caret at the end of the paragraph "SuperEditor is a toolkit..."
3. Press the new line button
4. Press the button to convert the paragraph to a hr

The editor crashes with the following exception: 

```
I/ViewRootImpl@6183b6c[MainActivity](29157): ViewPostIme pointer 0
I/ViewRootImpl@6183b6c[MainActivity](29157): ViewPostIme pointer 1
I/IMM_LC  (29157): showSoftInput(View,I)
I/IMM_LC  (29157): ssi() - flag : 0 view : com.supereditor.example reason = SHOW_SOFT_INPUT
I/IMM_LC  (29157): ssi() view is not EditText
E/flutter (29157): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Exception: Couldn't find styler to create component for document node: TextNode
E/flutter (29157): #0      SingleColumnLayoutPresenter._createNewViewModel
E/flutter (29157): #1      SingleColumnLayoutPresenter.updateViewModel
E/flutter (29157): #2      _SingleColumnDocumentLayoutState._onPresenterMarkedDirty
E/flutter (29157): #3      SingleColumnLayoutPresenterChangeListener.onPresenterMarkedDirty
E/flutter (29157): #4      SingleColumnLayoutPresenter._onDocumentChange
E/flutter (29157): #5      MutableDocument.onTransactionEnd
E/flutter (29157): #6      Editor.execute
E/flutter (29157): #7      KeyboardEditingToolbar._convertToHr
E/flutter (29157): #8      _InkResponseState.handleTap
E/flutter (29157): #9      GestureRecognizer.invokeCallback
E/flutter (29157): #10     TapGestureRecognizer.handleTapUp
E/flutter (29157): #11     BaseTapGestureRecognizer._checkUp
E/flutter (29157): #12     BaseTapGestureRecognizer.handlePrimaryPointer
E/flutter (29157): #13     PrimaryPointerGestureRecognizer.handleEvent
E/flutter (29157): #14     PointerRouter._dispatch
E/flutter (29157): #15     PointerRouter._dispatchEventToRoutes.<anonymous closure>
E/flutter (29157): #16     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
E/flutter (29157): #17     PointerRouter._dispatchEventToRoutes
E/flutter (29157): #18     PointerRouter.route
E/flutter (29157): #19     GestureBinding.handleEvent
E/flutter (29157): #20     GestureBinding.dispatchEvent
E/flutter (29157): #21     RendererBinding.dispatchEvent
E/flutter (29157): #22     GestureBinding._handlePointerEventImmediately
E/flutter (29157): #23     GestureBinding.handlePointerEvent
E/flutter (29157): #24     GestureBinding._flushPointerEventQueue
E/flutter (29157): #25     GestureBinding._handlePointerDataPacket
E/flutter (29157): #26     _invoke1 (dart:ui/hooks.dart:158:13)
E/flutter (29157): #27     PlatformDispatcher._dispatchPointerDataPacket (dart:ui/platform_dispatcher.dart:382:7)
E/flutter (29157): #28     _dispatchPointerDataPacket (dart:ui/hooks.dart:91:31)
E/flutter (29157):
D/InsetsController(29157): show(ime(), fromIme=true)
```

angelosilvestre commented:
Steps to reproduce:

1. Place the caret at the list item "SuperEditor| is a ready..."
2. Press ctrl + v to paste some text

The editor crashes with the following exception:

```
Restarted application in 588ms.
flutter: Initializing logger: ExampleApp
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Exception: Can't handle pasting text within node of type: [TextNode] - text: [AttributedText] - "SuperEditor is a ready-made, configurable document editing experience."
[AttributedSpans] (0 spans):, metadata: {blockType: [NamedAttribution]: listItem}
#0      PasteEditorCommand.execute
#1      _DocumentEditorCommandExecutor.executeCommand
#2      Editor._executeCommand
#3      Editor.execute
#4      CommonEditorOperations._paste
<asynchronous suspension>
```

matthew-carroll commented on super_editor/lib/src/core/document_composer.dart:
```diff
@@ -240,3 +441,84 @@ class SelectionReason {
+
+  @override
+  void execute(EditContext context, CommandExecutor executor) {
+    context.find<MutableDocumentComposer>(Editor.composerKey)._composingRegion.value = composingRegion;
```
I think it's OK. This command is tied to our specific composer implementation. Other editors would likely need their own composer, and their own composer commands.

matthew-carroll commented on super_editor/lib/src/default_editor/document_hardware_keyboard/document_keyboard_actions.dart:
```diff
@@ -245,13 +254,9 @@ ExecutionInstruction backspaceToRemoveUpstreamContent({
+  required SuperEditorContext editContext,
   required RawKeyEvent keyEvent,
 }) {
-  if (keyEvent is! RawKeyDownEvent) {
-    return ExecutionInstruction.continueExecution;
-  }
```
I don't remember why that was deleted. I guess I'll bring it back and we'll see if something breaks.

matthew-carroll commented on super_editor/lib/src/default_editor/multi_node_editing.dart:
```diff
@@ -2,39 +2,423 @@ import 'dart:math';
+
+    executor.logChanges([
+      DocumentEdit(
+        NodeInsertedEvent(newNode.id, document.getNodeIndexById(newNode.id)),
```
What do you mean by "use newNode"?

matthew-carroll replied:
Oh, you mean pass it as the 2nd argument? Yeah, I think so.

matthew-carroll replied:
Actually, no, the 2nd argument is the index of the new node. We don't have the index at that point.

raulmabe-labhouse commented:
I noticed a bug on this branch

### Expected behavior
When tapping on a `DocumentPosition`, the selection updates accordingly. However, when tapping again on the same `DocumentPosition`, it should show the toolbar.

### Actual behavior
When tapping on a `DocumentPosition`, the selection updates accordingly. However, when tapping again on the same `DocumentPosition`, it does not show the toolbar.

### Cause
This works well on `stable` branch, but because of now selection changes are part of a document change, it triggers the listener `_onDocumentChange` in `document_gestures_touch_ios.dart`  which hides the toolbar just after trying to show it on `_onTapUp`

matthew-carroll commented:
Thanks @raulmabe-labhouse - that issue should now be resolved.

matthew-carroll commented:
I rebased against `main` to bring in a refactor of the delta editing system, which fixed a newline problem on Android.

miguelcmedeiros commented on super_editor/lib/src/default_editor/multi_node_editing.dart:
```diff
@@ -2,39 +2,423 @@ import 'dart:math';
+      // the target node.
+      nodeMoveEvents.add(
+        DocumentEdit(
+          NodeMovedEvent(nodeId: document.getNodeAt(i)!.id, from: i, to: i - 1),
```
While debugging and issue on our side with moving nodes, I noticed that we are decrementing instead of incrementing the `to` index. When moving a node to index 0, it would define `to` as -1.
```suggestion 
          NodeMovedEvent(nodeId: document.getNodeAt(i)!.id, from: i, to: i + 1),
```

matthew-carroll replied:
I think the original behavior is correct, and I don't think it should be possible for this situation to impact the 0'th node. Did you run into that in real execution?

With the 0'th node being the very first node in the list, the only way the 0'th node would be impacted by this loop is if that's the node that's being moved somewhere else. If it's being moved somewhere else, then it followed the if-statement above, rather than the adjustment you commented on.

When a node moves to a new location in the list, it causes all other nodes between those two positions to move up by one, hence the `- 1`.

miguelcmedeiros replied:
I wrote a small test where we can check this:
```dart
import 'package:flutter_test/flutter_test.dart';
import 'package:super_editor/super_editor.dart';

void main() {
  test('move a node', () {
    final document = MutableDocument(
      nodes: [
        ParagraphNode(
          id: 'p1',
          text: AttributedText(text: 'Paragraph 1'),
        ),
        ParagraphNode(
          id: 'p2',
          text: AttributedText(text: 'Paragraph 2'),
        ),
        ParagraphNode(
          id: 'p3',
          text: AttributedText(text: 'Paragraph 3'),
        ),
      ],
    );
    final editor = Editor(
      editables: {Editor.documentKey: document},
      requestHandlers: [
        (request) => request is MoveNodeRequest
            ? MoveNodeCommand(
                nodeId: request.nodeId,
                newIndex: request.newIndex,
              )
            : null,
      ],
    );
    const request = MoveNodeRequest(
      nodeId: 'p3',
      newIndex: 0,
    );
    final changeLog = <DocumentChangeLog>[];
    document.addListener(changeLog.add);

    editor.execute([request]);

    expect(
      document.nodes.map((node) => node.id),
      ['p3', 'p1', 'p2'],
    );

    print(changeLog.map((e) => e.changes));
    expect(changeLog.length, 1);
    expect(
      changeLog.first.changes,
      [
        const NodeMovedEvent(nodeId: 'p1', from: 0, to: 1),
        const NodeMovedEvent(nodeId: 'p2', from: 1, to: 2),
        const NodeMovedEvent(nodeId: 'p3', from: 2, to: 0),
      ],
    );
  });
}

```

The last `expect` where we check the change log fails. The changes reported with the current code are:
```dart
    expect(
      changeLog.first.changes,
      [
        const NodeMovedEvent(nodeId: 'p1', from: 0, to: -1),
        const NodeMovedEvent(nodeId: 'p2', from: 1, to: 0),
        const NodeMovedEvent(nodeId: 'p3', from: 2, to: 0),
      ],
    );
```

miguelcmedeiros commented on super_editor/lib/src/default_editor/multi_node_editing.dart:
```diff
@@ -2,39 +2,423 @@ import 'dart:math';
+    document.moveNode(nodeId: nodeId, targetIndex: newIndex);
+
+    // Report all the node movements.
+    executor.logChanges(nodeMoveEvents);
```
Even considering the fix mentioned [here](https://github.com/superlistapp/super_editor/pull/1130/files#r1209341121), it seems that the document is not being rebuilt (i.e. the nodes remain on the same positions) after these changes are logged.
If we do a hot reload, then the document is rebuilt and the nodes are placed where expected.

matthew-carroll replied:
Is there a demo in the example app that reproduces the issue? I'm not sure how to verify a fix without a repro.

miguelcmedeiros replied:
There is no example in the demo app at the moment. This request/command pair is not used in `super_editor` as far as I can tell.

I tried it out in the demo app by hacking into `lib/src/default_editor/tasks.dart` and changing the [`setComplete`](https://github.com/superlistapp/super_editor/blob/1125_editor-pipeline-test-suite/super_editor/lib/src/default_editor/tasks.dart#L89) callback into:
```dart
      setComplete: (bool isComplete) {
        _editor.execute([
          ChangeTaskCompletionRequest(
            nodeId: node.id,
            isComplete: isComplete,
          ),
          MoveNodeRequest(nodeId: node.id, newIndex: 0),
        ]);
      },
```

When running the demo app, it should move whatever task we mark as complete/incomplete to the top of the document.

miguelcmedeiros commented:
@matthew-carroll 
I'm not able to open a link ("linkified" as the user types). It seems that it misses the schema (e.g. http://) and the url_launcher fails to open.

matthew-carroll commented:
> @matthew-carroll I'm not able to open a link ("linkified" as the user types). It seems that it misses the schema (e.g. http://) and the url_launcher fails to open.

I pushed a change that automatically adds the scheme if it's not provided. Let me know if you have any more issues with that.

miguelcmedeiros commented:
> I pushed a change that automatically adds the scheme if it's not provided. Let me know if you have any more issues with that.

Works 🙌

matthew-carroll commented:
@miguelcmedeiros I pushed a fix for node re-ordering.

There was already a test to verify that we can move a node from a lower index to a higher index. The problem was specifically with moving a node from a higher index to a lower index. I added a test for that, and fixed the failure.

matthew-carroll commented:
Update @miguelcmedeiros @Jethro87 @jmatth @raulmabe-labhouse - It looks like we've gotten to the point where issues are trickling in at a slow rate. At some point soon I'll have to call it and just merge. If there are any more fundamental concerns or issues, now is the least expensive time to raise them.

Aside from API improvements that we can make along the way, are there any fundamental or significant concerns, objections, or issues that you have with this approach?

Does this approach reflect what you would expect from an atomic command/pipeline system?

Is there anything else you'd like to work through before we commit to this direction and begin the process of incrementally improving the API?

miguelcmedeiros commented:
> Aside from API improvements that we can make along the way, are there any fundamental or significant concerns, objections, or issues that you have with this approach?
> 
> Does this approach reflect what you would expect from an atomic command/pipeline system?
> 
> Is there anything else you'd like to work through before we commit to this direction and begin the process of incrementally improving the API?

Although it was a big endeavor to migrate our app to support atomic commands, I appreciate how the APIs are designed and it reflects a clearer separation of concerns.

matthew-carroll commented:
@miguelcmedeiros can you try again for the layout component re-ordering? I solved a problem in the view model diff'ing in the presenter, but I don't know if that's the final problem from a UI perspective.

matthew-carroll commented:
@miguelcmedeiros I pushed another fix for moving nodes. Previously, we had no representation of a moved components in the view model change listener. We only had add, remove, and change. I added the concept of moving components and I hooked up the layout to re-flow all the components whenever any of them move.

I'm sure there's a more efficient way to invalidate the layout, but I think this should solve your problem, and hopefully it won't be too hard on performance.

Jethro87 commented:
@matthew-carroll I came across this error when trying to insert an image in a non-empty node:

```
════════ Exception caught by foundation library ════════════════════════════════
The following _TypeError was thrown while dispatching notifications for PausableValueNotifier<DocumentSelection?>:
type 'UpstreamDownstreamNodePosition' is not a subtype of type 'TextPosition' in type cast

When the exception was thrown, this was the stack
#0      SuperEditorState._updateComposerPreferencesAtSelection
super_editor.dart:467
#1      ChangeNotifier.notifyListeners
change_notifier.dart:403
#2      ValueNotifier.value=
change_notifier.dart:530
#3      PausableValueNotifier.resumeNotifications
pausable_value_notifier.dart:49
#4      MutableDocumentComposer.onTransactionEnd
document_composer.dart:144
#5      Editor.execute
editor.dart:164
#6      CommonEditorOperations._insertBlockLevelContent
common_editor_operations.dart:1787
#7      CommonEditorOperations.insertImage
common_editor_operations.dart:1711
#8      _InsertImageFormattingActionState._insertMultipleImages
insert_image_action.dart:234
<asynchronous suspension>
#9      _InsertImageFormattingActionState._pickDesktopOrWebImage
insert_image_action.dart:183
<asynchronous suspension>
#10     _InsertImageFormattingActionState._pickImage
insert_image_action.dart:104
<asynchronous suspension>
The PausableValueNotifier<DocumentSelection?> sending notification was: PausableValueNotifier<DocumentSelection?>#8d474([DocumentSelection] - 
  base: ([DocumentPosition] - node: "b1be4669-a719-432d-aea7-28e0db0af5cd", position: (TextPosition(offset: 5, affinity: TextAffinity.upstream))),
  extent: ([DocumentPosition] - node: "b1be4669-a719-432d-aea7-28e0db0af5cd", position: (TextPosition(offset: 5, affinity: TextAffinity.upstream))))
════════════════════════════════════════════════════════════════════════════════

════════ Exception caught by foundation library ════════════════════════════════
Exception: Expected a TextNodePosition for position1 but received a UpstreamDownstreamNodePosition
════════════════════════════════════════════════════════════════════════════════
```

For example, inserting at the end or middle of "hello" rather than an empty node.

matthew-carroll commented on super_editor/lib/src/default_editor/multi_node_editing.dart:
```diff
@@ -2,39 +2,434 @@ import 'dart:math';
+    }
+
+    // Move the target node to its destination index.
+    document.moveNode(nodeId: nodeId, targetIndex: newIndex);
```
Did you hit a bug, or did you read the implementation and assume it's a problem?

There's an existing test which moves a lower index node to a higher index. That test verifies all final node positions, and the reported change events.

Also, here's a run through of the implementation:

["1", "2", "3", "4"]

Move node with "1" to index 2 (where the "3" node appears)

Remove node: ["1"], ["2", "3", "4"]

Insert node at index 2: ["2", "3", "1", "4"]

Am I missing anything?

miguelcmedeiros replied:
I hit the bug when in our app, but turns out that this was indeed a bug on our side.

matthew-carroll commented:
@Jethro87 is there a way to repro that in the example app? Also, which platform are you using?

Jethro87 commented:
@matthew-carroll The example app doesn't include image functionality, but it can be hooked up easily enough. Any of the toolbar buttons can call `commonOps.insertImage(filePath)`, where `commonOps` is an instance of `CommonEditorOperations`.

I've tested this on macOS and Android so far.

matthew-carroll commented:
@Jethro87 I'm unable to trigger this error when running the example app on Mac.

I tried using `CommonEditorOperations` as well as directly submitting an `InsertNodeAtCaretRequest`. I also tried running this with a collapsed selection and an expanded selection.

I tried inserting the image in a paragraph, and in a header.

All attempts worked.


https://github.com/superlistapp/super_editor/assets/7259036/3f9bc1c2-2b67-4e8c-a5c3-98b154d188a3

matthew-carroll commented:
@miguelcmedeiros that issue should now be fixed.

What was happening was that CMD was triggering the enabling and disabling of interaction mode (to tap on links). That activation and deactivation goes through the editor pipeline, like any other request. This revealed a couple of things to improve.

First, the `DocumentComposer` pauses and restarts selection activations while the editor pipeline is processing changes. But when the `DocumentComposer` reactivated its notifications, it automatically sent one of those selection change notifications, even if the selection didn't change during that transaction. I've fixed this and added a test.

Second, the desktop gesture system was assuming that any time a selection change notification came in, it definitely meant that the selection changed, in which case it should auto-scroll. That was a reasonable assumption, but it's unnecessarily fragile. I changed that behavior to check the previous selection vs the current selection. I also added a test.

matthew-carroll commented:
I'm going to merge in this PR now that we've had ample time to locate and resolve bugs.

I'm going to immediately follow up with some content conversion and recognition features, like markdown headers and user tagging.
